### PR TITLE
Removed outputOnly from list of languages

### PR DESCRIPTION
**Pull request by antedotee**
Removed output only language from dropdown bar in languages to select

# Description

Earlier
CleanShot 2024-03-17 at 01 13 21@2x
Now

CleanShot 2024-03-17 at 01 15 04@2x

Fixes: #4294 

# Checklist:

- [x] The code follows the [coding guidelines](https://github.com/omegaup/omegaup/wiki/Coding-guidelines) of omegaUp.
- [x] The tests were executed and all of them passed.
- [ ] If you are creating a feature, the new tests were added.
- [ ] If the change is large (> 200 lines), this PR was split into various Pull Requests. It's preferred to create one PR for changes in controllers + unit tests in PHPUnit,  and then another Pull Request for UI + tests in Jest, Cypress or both.

### CODE REVIEW

**Review by pabo99 — APPROVED** [review]
In the future, we aim to use a unified list of languages, which includes this list. @HNOONa-0  and I are currently exploring the best approach to achieve this.

For now, this change is acceptable!

**Comment by pabo99** [comment]
Could you please update your branch? This may resolve the failure in the Python test.

**Comment by antedotee** [comment]
done

File: frontend/server/src/Controllers/Run.php [hunk 1]
```diff
@@ -34,7 +34,6 @@ class Run extends \OmegaUp\Controllers\Controller {
         'rb' => 'Ruby (2.7)',
         'cs' => 'C# (10, dotnet 6.0)',
         'pas' => 'Pascal (fpc 3.0)',
-        'cat' => 'Output Only',
```
Comment (pabo99) [inline frontend/server/src/Controllers/Run.php:37]: Ok, it appears that this language is being used in a test, which is causing the test to fail. Therefore, simply removing this line is not going to be the solution.

We should only hide this option in the user's profile. To achieve this, in `Controllers/User.php`, you should add a function named `getSupportedLanguages`:

```php

    private static function getSuportedProgrammingLanguages() {
        return array_filter(
            \OmegaUp\Controllers\Run::SUPPORTED_LANGUAGES,
            fn($key) => $key !== 'cat',
            ARRAY_FILTER_USE_KEY
        );
    }
``` 

Next, replace all occurrences of `\OmegaUp\Controllers\Run::SUPPORTED_LANGUAGES` in this file with calls to the new `getSupportedLanguages` function, which filters out the 'cat' language.

**Review by pabo99 — COMMENTED** [review]
I've left you a suggestion. 

Also, could you please add unit tests in PHP to verify the new behavior? You might consider using an existing test like `testGetProfileDetails`, as it already retrieves the list of supported languages.

**Comment by heduenas** [comment]
@antedotee this change in its current form will also delete output only from the list of languages when making a submission or when creating a contest/course. We want to eliminate it only from the list of preferred languages.

File: frontend/tests/controllers/UserProfileTest.php [hunk 2]
```diff
@@ -678,6 +678,40 @@ public function testGetProfileDetails(
...
+        'rb' => 'Ruby (2.7)',
+        'cs' => 'C# (10, dotnet 6.0)',
+        'pas' => 'Pascal (fpc 3.0)',
+        'cat' => 'Output Only',
+        'hs' => 'Haskell (ghc 8.8)',
+        'lua' => 'Lua (5.3)',
+        'go' => 'Go (1.18.beta2)',
+        'rs' => 'Rust (1.56.1)',
+        'js' => 'JavaScript (Node.js 16)',
+    ];
```
[the remark this answers is not in the document]
Comment (antedotee) [reply frontend/tests/controllers/UserProfileTest.php]: yes,i changed that😅

File: frontend/server/src/Controllers/Contest.php [hunk 3]
```diff
@@ -1324,7 +1324,7 @@ public static function getContestNewForTypeScript(
         return [
             'templateProperties' => [
                 'payload' => [
-                    'languages' => \OmegaUp\Controllers\Run::SUPPORTED_LANGUAGES,
+                    'languages' => \OmegaUp\Controllers\User::getSuportedProgrammingLanguages(),
```
[the remark this answers is not in the document]
Comment (antedotee) [reply frontend/server/src/Controllers/Contest.php]: yes,sorry for mistake from my side🥲.I am reverting these changes now.

File: frontend/tests/controllers/UserProfileTest.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (antedotee) [reply frontend/tests/controllers/UserProfileTest.php]: did the changes as you said🙏.waiting for your suggestions of the test.

**Review by heduenas — APPROVED** [review]

**Review by pabo99 — COMMENTED** [review]
Almost done.

You need to test that the language 'cat' doesn't exist when the user gets the `programming_languages` array in their profile details.

In the file `UserProfileTest`, we have a test called `testGetProfileDetails`. This test calls the function `getProfileDetailsForTypeScript` and saves the user profile response in a variable called `$profile`.

At the end of the test, you could add the following assertion:
```php
$this->assertArrayNotHasKey('cat', $profile['programming_languages']);
``` 
and that's it!

Well, one more thing, if you could restore the files:

- `frontend/server/src/Controllers/Course.php`
- `frontend/tests/controllers/RunCreateTest.php`

because the changes made to those files are not needed.

File: frontend/server/src/Controllers/User.php [hunk 4]
```diff
@@ -3703,7 +3703,13 @@ private static function getSelectedFilter(
...
         return ['filteredBy' => null, 'value' => null];
     }
-
+    public static function getSupportedProgrammingLanguages(): array {
+        return array_filter(
+            \OmegaUp\Controllers\Run::SUPPORTED_LANGUAGES,
+            fn($key) => $key !== 'cat',
+            ARRAY_FILTER_USE_KEY
+        );
+    }
```
Comment (pabo99) [inline frontend/server/src/Controllers/User.php:3712]: For PHP, the array type declaration you added is correct, but let's remember that PHP is a weakly typed language. So, to have greater control over inputs and outputs in functions, we use a tool to add types. This tool is PSALM, so we should be more specific when declaring the type. In this case, it is an array with string keys and string values, which is declared as follows:

```suggestion

    /**
     * @return array<string,string>
     */
    public static function getSupportedProgrammingLanguages(): array {
        return array_filter(
            \OmegaUp\Controllers\Run::SUPPORTED_LANGUAGES,
            fn($key) => $key !== 'cat',
            ARRAY_FILTER_USE_KEY
        );
    }

```

**Review by pabo99 — COMMENTED** [review]
One last comment!

**Pull request merged**